tox.ini

```diff
@@ -4,12 +4,12 @@ envlist =
     {3.10,3.11,3.12,pypy310}-main
     {3.10,3.11,3.12,pypy310}-5.0.x
     {3.8,3.9,3.10,3.11,3.12,pypy38,pypy39,pypy310}-4.2.x
-    {3.8,3.9,3.10,pypy38,pypy39,pypy310}-3.2.x
```

Comment:
I was able to get this working in a branch where I added typing annotations. I'll see if I can uplift that fix to this PR.

Reply:
I haven't figured out the changes that made these work yet. It seems unlikely that `pypy` changes behavior based on types, so I'll work to tease apart the code changes from the typing changes and see if that fixes it.

I think this PR is still mergable because of the improvement in the testing configuration.